File: SortingAlgorithms/InsertionSort/MergeSort/Mergesort.js
export class MergeSort {
  static sortAscending(arr) {
    this.sortAscendingInPlace(arr, 0, arr.length);
  }
  static sortDecending(arr) {
    this.sortDecendingInPlace(arr, 0, arr.length);
  }
  static sortAscendingArr(arr) {
    if (arr.length == 1) {
      return arr;
    }

    let left = this.sortAscendingArr(arr.slice(0, arr.length / 2));
    let right = this.sortAscendingArr(arr.slice(arr.length / 2, arr.length));
    return this.mergeAscendingArr(left, right);
  }

  static mergeAscendingArr(left, right) {
    let joinedArr = [];
    let i = 0,
      j = 0;
    while (i < left.length && j < right.length) {
      if (left[i] < right[j]) {
        joinedArr.push(left[i++]);
      } else {
        joinedArr.push(right[j++]);
      }
    }
    while (i < left.length) {
      joinedArr.push(left[i++]);
    }
    while (j < right.length) {
      joinedArr.push(right[j++]);
    }
    return joinedArr;
  }
  static sortAscendingInPlace(arr, start, end) {
    if (end - start == 1) {
      return;
    }
    let mid = Math.floor((start + end) / 2);

    this.sortAscendingInPlace(arr, start, mid);
    this.sortAscendingInPlace(arr, mid, end);
    return this.mergeAscendingInPlace(arr, start, mid, end);
  }

  static mergeAscendingInPlace(arr, start, mid, end) {
    let joinedArr = [];
    let i = start,
      j = mid;
    while (i < mid && j < end) {
      if (arr[i] < arr[j]) {
        joinedArr.push(arr[i++]);
      } else {
        joinedArr.push(arr[j++]);
      }
    }
    while (i < mid) {
      joinedArr.push(arr[i++]);
    }
    while (j < end) {
      joinedArr.push(arr[j++]);
    }

    for (let k = 0; k < joinedArr.length; k++) {
      arr[start + k] = joinedArr[k];
    }
  }
  static sortDecendingArr(arr) {
    if (arr.length == 1) {
      return arr;
    }

    let left = this.sortDecendingArr(arr.slice(0, arr.length / 2));
    let right = this.sortDecendingArr(arr.slice(arr.length / 2, arr.length));
    return this.mergeDecendingArr(left, right);
  }

  static mergeDecendingArr(left, right) {
    let joinedArr = [];
    let i = 0,
      j = 0;
    while (i < left.length && j < right.length) {
      if (left[i] > right[j]) {
        joinedArr.push(left[i++]);
      } else {
        joinedArr.push(right[j++]);
      }
    }
    while (i < left.length) {
      joinedArr.push(left[i++]);
    }
    while (j < right.length) {
      joinedArr.push(right[j++]);
    }
    return joinedArr;
  }
  static sortDecendingInPlace(arr, start, end) {
    if (end - start == 1) {
      return;
    }
    let mid = Math.floor((start + end) / 2);

    this.sortDecendingInPlace(arr, start, mid);
    this.sortDecendingInPlace(arr, mid, end);
    return this.mergeDecendingInPlace(arr, start, mid, end);
  }

  static mergeDecendingInPlace(arr, start, mid, end) {
    let joinedArr = [];
    let i = start,
      j = mid;
    while (i < mid && j < end) {
      if (arr[i] > arr[j]) {
        joinedArr.push(arr[i++]);
      } else {
        joinedArr.push(arr[j++]);
      }
    }
    while (i < mid) {
      joinedArr.push(arr[i++]);
    }
    while (j < end) {
      joinedArr.push(arr[j++]);
    }

    for (let k = 0; k < joinedArr.length; k++) {
      arr[start + k] = joinedArr[k];
    }
  }
}

File: SortingAlgorithms/InsertionSort/MergeSort/MergeSort.java
public class MergeSort {
	public static void sortAscending(int arr[]) {
		sortAscendingInPlace(arr, 0, arr.length);
	}
	public static void sortDecending(int arr[]) {
		sortDecendingInPlace(arr, 0, arr.length);
	}
	public static int[] sortAscendingArr(int arr[]) {

		if (arr.length == 1) {
			return arr;
		}

		int left[] = sortAscendingArr(Arrays.copyOfRange(arr, 0, arr.length / 2));
		int right[] = sortAscendingArr(Arrays.copyOfRange(arr, arr.length / 2, arr.length));

		return mergeAscendingArr(left, right);
	}

	private static int[] mergeAscendingArr(int left[], int right[]) {
		int arr[] = new int[left.length + right.length];
		int i = 0, j = 0, k = 0;

		while (i < left.length && j < right.length) {
			if (left[i] < right[j]) {
				arr[k++] = left[i++];
			} else {
				arr[k++] = right[j++];
			}
		}

		while (i < left.length) {
			arr[k++] = left[i++];
		}
		while (j < right.length) {
			arr[k++] = right[j++];
		}

		return arr;
	}

	public static void sortAscendingInPlace(int arr[], int start, int end) {

		if (end - start == 1) {
			return;
		}
		int mid = (start + end) / 2;

		sortAscendingInPlace(arr, start, mid);
		sortAscendingInPlace(arr, mid, end);

		mergeAscendingInPlace(arr, start, mid, end);
	}

	private static void mergeAscendingInPlace(int[] arr, int start, int mid, int end) {
		int joined[] = new int[end - start];
		int i = start, j = mid, k = 0;

		while (i < mid && j < end) {
			if (arr[i] < arr[j]) {
				joined[k++] = arr[i++];
			} else {
				joined[k++] = arr[j++];
			}
		}

		while (i < mid) {
			joined[k++] = arr[i++];
		}
		while (j < end) {
			joined[k++] = arr[j++];
		}
		for (k = 0; k < joined.length; k++) {
			arr[start+k]=joined[k];
			
		}
	}

	public static int[] sortDecendingArr(int[] arr) {
		if (arr.length == 1) {
			return arr;
		}

		int left[] = sortDecendingArr(Arrays.copyOfRange(arr, 0, arr.length / 2));
		int right[] = sortDecendingArr(Arrays.copyOfRange(arr, arr.length / 2, arr.length));

		return mergeDecendingArr(left, right);

	}

	private static int[] mergeDecendingArr(int[] left, int[] right) {
		int joined[] = new int[left.length + right.length];

		int i = 0, j = 0, k = 0;
		while (i < left.length && j < right.length) {
			if (left[i] > right[j]) {
				joined[k++] = left[i++];
			} else {
				joined[k++] = right[j++];
			}
		}
		while (i < left.length) {
			joined[k++] = left[i++];
		}

		while (j < right.length) {
			joined[k++] = right[j++];
		}
		return joined;

	}
	public static void sortDecendingInPlace(int arr[], int start, int end) {

		if (end - start == 1) {
			return;
		}
		int mid = (start + end) / 2;

		sortDecendingInPlace(arr, start, mid);
		sortDecendingInPlace(arr, mid, end);

		mergeDecendingInPlace(arr, start, mid, end);
	}

	private static void mergeDecendingInPlace(int[] arr, int start, int mid, int end) {
		int joined[] = new int[end - start];
		int i = start, j = mid, k = 0;

		while (i < mid && j < end) {
			if (arr[i] > arr[j]) {
				joined[k++] = arr[i++];
			} else {
				joined[k++] = arr[j++];
			}
		}

		while (i < mid) {
			joined[k++] = arr[i++];
		}
		while (j < end) {
			joined[k++] = arr[j++];
		}
		for (k = 0; k < joined.length; k++) {
			arr[start+k]=joined[k];
			
		}
	}

}